fix: FlexiblePathfinding migration follow-ups

skaldarnar opened the pull request:
- style: format `.behavior` files
- chore: add TODO ideas and questions in several places
- chore: add some debug logging
- feature/fix: small adjustments to Behaviors logic

Follow-up PR to Terasology/Behaviors#89.

Related PRs (extracted from this): Terasology/Behaviors#94, Terasology/Behaviors#96, Terasology/Behaviors#97, Terasology/Behaviors#99.

Contributes to MovingBlocks/Terasology#4981.

Depends on Terasology/FlexiblePathfinding#26 and Terasology/FlexiblePathfinding#27.

skaldarnar committed:
chore: format (some) behavior files

Formatted `critter`, `doRandomMove`, `hostileCritter`, and `stray` behavior files with indent of 2 spaces.

skaldarnar committed:
fix?: use plain `selector` instead of `dynamic` for critter behavior

The structure of having a guarded behavior lookup (`flee`) and a fallback behavior lookup (`stray`) to me looks like we're good with running nodes until we find the first one that is successful.

skaldarnar committed:
refactor: make CheckAttackStopAction more comprehensible

Applies simple refactorings such as renaming or inversion of return values to make the code defining the action more readable.

In addition, added a comment to discuss whether this should be a (complex) action or rather a behavior.

skaldarnar committed:
fix: `@In` annotation not working in SetTargetToNearbyBlockNode

In addition, adds a couple of debug logs to investigate the control flow when running the action.

These debug lines revealed that exceptions during the execution of an action are treated as failed actions. The behavior tree runner logs a warning, but it is easy to miss...

skaldarnar committed:
chore: formatting and code style improvements

jdrueckert committed:
feat: use pathfinding-based naiveMoveTo instead of direct move_to

jdrueckert committed:
fix: write target into MinionMoveComponent.pathGoal required by find_path action

jdrueckert committed:
fix: mark behavior running while pathfinding is not completed yet

skaldarnar committed:
docs: improve documentation and logging

Annotate methods of MinionMoveComponent with Javadoc (and questions) to understand the internals of that component better.

Improve logging of MoveToAction and SetTargetToNearbyBlockNode.

skaldarnar committed:
fix: workaround for `@In` in SetTargetToNearbyBlockAwayFromInstigatorAction

skaldarnar committed:
chore: use non-deprecated constructor for CharacterMoveInputEvent

skaldarnar committed:
docs: add more Javadoc and TODOs to MinionMoveComponent

jdrueckert commented:
I just noticed that the now pathfinding-based `doRandomMove` works pretty okay (including moving uphill).
However, it (1) does time out rather frequently when actually moving along the path or (2) the randomly chosen target block just "doesn't make sense".

For (1) we need to investigate why it times out after having found what appears to be a valid path (and judging from the debug path visualization actually is a reasonable path).
For (2) I think what we need here would be some more movement-type-based restrictions on what a "reasonable random target block" is, which should be reflected in the plugin's `isReachable` method.
i.e. for an entity that is walking or leaping, any target position that does not have a solid block underneath doesn't make sense (even if gravity does not apply to mobs if they don't move - btw I still feel like it does when they jump, but I don't know why).
on the other hand for an entity that is flying or swimming, that's totally fine.
for swimming however any block that is not liquid is not a reasonable target.

skaldarnar committed:
fix: only render debug overlay when debug mode is enabled

skaldarnar committed:
chore: format lure behavior

skaldarnar committed:
debug: add more logging to behaviors and actions

- add log messages when entering a behavior: critter (root), doRandomMove, naiveMoveTo, stray

skaldarnar committed:
debug: change logging format of LogAction

Now prints out messages in the format:

```
17:43:24.551 [main] DEBUG o.t.m.behaviors.actions.LogAction - Actor 2502: ---- critter ----
```

skaldarnar committed:
debug: debug log messages for various Actions

skaldarnar committed:
Revert "fix?: use plain `selector` instead of `dynamic` for critter behavior"

This reverts commit f4444b3a6faead9f48be90318c2651174e5bd6d9.

skaldarnar committed:
fix: use FLYING movement mode in Walking and Leaping

The pathfinding plugins assume that entities are not affected by gravity. To match this assumption, we chose a movement mode that neglects gravity, but works similar to the regular walking mode: FLYING.

skaldarnar committed:
fix(set_target_nearby_block): ensure target block is walkable

Without this check, we can end up with target blocks that are not walkable, screwing up the pathfinding logic. This small addition will fail the behavior action when the target is not walkable, in the hope that we'll just give it another try.

Several ideas on how to further improve this behavior action are added as TODO comments.

skaldarnar committed:
feat: add FallingMovementPlugin

This is a special movement plugin that will enable gravity on entities while they are using the falling movement mode.

skaldarnar committed:
debug: add more debug logs to movement plugins

jdrueckert commented:
![](https://github.trello.services/images/mini-trello-icon.png) [FlexibleMovement as a replacement for PathFinding](https://trello.com/c/AGtBwude/189-flexiblemovement-as-a-replacement-for-pathfinding)

jdrueckert committed:
Merge branch 'develop' into fix/behavior-follow-ups

jdrueckert committed:
feat: optimize findRandomNearbyBlock

- only accept candidate if it increases the distance to the start block

jdrueckert committed:
Merge remote-tracking branch 'origin/fix/behavior-follow-ups' into fix/behavior-follow-ups

jdrueckert committed:
chore: clean up comment

jdrueckert commented:
I implemented the optimization proposed in the todo comment for `findRandomNearbyBlock()` in `SetTargetToNearbyBlockNode.java` by checking for the candidate increasing the distance to the start block in addition to it being reachable. I put it before the reachability check because I think it happens more often that a candidate would be reachable but does not "get us forward" - I'm fine with switching the order if somebody thinks it's the other way round or roughly equally likely but checking reachability is cheaper.

Also, after observing our deer a bit more, I think that something about the moveTo completion check is off. I noticed that https://github.com/Terasology/Behaviors/blob/fix/behavior-follow-ups/src/main/java/org/terasology/module/behaviors/actions/MoveToAction.java#L72 does not result in `true` for instance for position (-73.63, 29.00, 40.99) and target (-73, 29, 41). Considering our NPCs do have a collision box and do not move as ghosts, I would deem this as close enough to tell moveTo that it's done. Currently we don't and try to get even close to our target until we time out eventually.

Creating a Vector3i from the Vector3f position with `RoundingMode.CEILING` (which I think we deemed to be the default to use? correct me if wrong) would match the target:
> `new Vector3i(new Vector3f(-73.63f, 29.00f, 40.99f), RoundingMode.CEILING)` --> (-73, 29, 41)

`Block.toBlockPos` [uses `TeraMath.floorToInt` after adding 0.5f](https://github.com/MovingBlocks/Terasology/blob/develop/engine/src/main/java/org/terasology/engine/world/block/Blocks.java#L63-L69) to every dimension of the vector, which does not match the target:
> `Blocks.toBlockPos(new Vector3f(-73.63f, 29, 40.99f))` --> (-74, 29, 41)

Do we think the `Blocks` utility class computation is correct? If so, should we introduce some kind of "tolerance" to the completion check in `moveToAction.java`?

skaldarnar committed:
fix: improve FallingMovementPlugin by using a direction for the movement

skaldarnar committed:
enable debug rendering again (reverts previous commit)

skaldarnar committed:
fix: allow for more movement cycles in MoveToAction

skaldarnar committed:
fix: use smaller target are to determine whether target is reached

skaldarnar commented:
While testing these changes we were wondering about the exact placement of Bullet physics colliders.
Since the entity just moves into the (intermediate) target block and then continues to the next target, a collision shape extending to the sides or below may very well prevent clean movement due to block collision. 

We were wondering how exactly the character colliders work, where they have their origin point, and how exactly their extents look in the world. Would be nice to have some debug rendering for this (in addition to the bounding boxes as defined by shape components).

We used the following change in the `BulletPhysics` class:


```diff
diff --git a/engine/src/main/java/org/terasology/engine/physics/bullet/BulletPhysics.java b/engine/src/main/java/org/terasology/engine/physics/bullet/BulletPhysics.java
index 61e802e76..513a810e6 100644
--- a/engine/src/main/java/org/terasology/engine/physics/bullet/BulletPhysics.java
+++ b/engine/src/main/java/org/terasology/engine/physics/bullet/BulletPhysics.java
@@ -566,6 +566,8 @@ private CharacterCollider createCharacterCollider(EntityRef owner) {
                     "CharacterMovementComponent.");
         }
         Vector3f pos = locComp.getWorldPosition(new Vector3f());
+        // adjust the position to be at the bottom center
+        pos.sub(0, movementComp.height / 2f, 0);
         final float worldScale = locComp.getWorldScale();
         final float height = (movementComp.height - 2 * movementComp.radius) * worldScale;
         final float width = movementComp.radius * worldScale;
```

skaldarnar committed:
test: fix MovementTests by moving `testCharacter` prefab to top-level

There seems to be some issue with loading prefabs from nested
sub-folders in tests. For some reason, the AssetManager could not find
the `Behaviors:testCharacter` prefab which was located at
```
assets/prefabs/test/testCharacter.prefab
```
To fix this (for now) I moved the asset one level up, which seems
sufficient for the engine to find the prefab also for test cases.

In addition, I've added CoreAssets as dependency for the MTE tests, as
they are referencing blocks from that module when setting up the test
world.

jdrueckert committed:
refactor: structure test cases into separate tests based on plugin

jdrueckert committed:
refactor(tests): extract leaping tests that only need walking

jdrueckert committed:
test: add walking test case "gap"

jdrueckert committed:
tests: add leaping test cases

jdrueckert committed:
test: add falling plugin test

jdrueckert committed:
test: add default plugin combination test for newly added tests

jdrueckert committed:
test: add test case for combined plugins

jdrueckert committed:
refactor: extract leaping test that requires falling

jdrueckert committed:
test: add tests for moving two steps up or down

jdrueckert committed:
test: add tests for moving one step up or down

jdrueckert commented:
* test case "up and down again" consistently fails
* the swimming-related "three dimensional movement" test case fails sporadically

jdrueckert committed:
chore: reorder test cases

jdrueckert committed:
test: add failureExpected parameter for implementing "negative" tests

jdrueckert committed:
test: remove large leaping steps test case

jdrueckert committed:
test: remove defaulted char height and radius params

jdrueckert committed:
test: remove unnecessary movement mode param for movement-categorized tests

jdrueckert committed:
test: add basic world and path definitions

skaldarnar committed:
feat: add "falling" to the default movement types

skaldarnar commented on src/test/java/org/terasology/module/behaviors/MovementTests.java:
```diff
@@ -182,6 +182,32 @@
                         new String[]{"walking", "leaping"}
+                ),
+                Arguments.of(
+                        "diagonally early up",
```
what do you mean by "early" and "late" here?

jdrueckert replied:
early in the sense that - assuming we don't actually walk diagonally but first to one side and then back into the diagonal - the first step brings us to the desired y-level and the second brings us on the same y-level to our target block
late vice versa means we first walk on the same y-level and then change y-levels to the desired one.

it's weird to explain and even weirder to name, so I'm happy about any naming help here

jdrueckert committed:
test: add basic test cases for various movement plugins

- and actually make use of failureExpected param

jdrueckert committed:
Merge remote-tracking branch 'origin/fix/behavior-follow-ups' into fix/behavior-follow-ups

skaldarnar commented on src/test/java/org/terasology/module/behaviors/MovementTests.java:
```diff
@@ -340,6 +1224,9 @@ private EntityRef createMovingCharacter(float height, float radius, Vector3i sta
 
         physicsEngine.removeCharacterCollider(entity);
         physicsEngine.getCharacterCollider(entity);
```
This could be improved by https://github.com/MovingBlocks/Terasology/pull/4996

jdrueckert committed:
fix: invert failureExpected param

jdrueckert committed:
chore: remove info logs

jdrueckert committed:
chore: destroy entity after test case run

jdrueckert committed:
fix: test case naming for combined movement tests

jdrueckert committed:
chore: reduce time out and remove purge between tests

- test worlds are set up with an initialization of the padded test world extents, so we can live without purging in-between

jdrueckert committed:
refactor: extract airHeight

skaldarnar committed:
test: destroy entity after test and remove code duplication

skaldarnar committed:
test: declare constants for test suite

skaldarnar committed:
test: use entity builder to create test character

jdrueckert committed:
chore: use more re-usable world definitions

jdrueckert committed:
Merge remote-tracking branch 'origin/fix/behavior-follow-ups' into fix/behavior-follow-ups

jdrueckert committed:
fix: expect "old" test cases to succeed

- also fixes some formatting

jdrueckert committed:
fix: add missing world for ascending late and descending early cases

skaldarnar committed:
test: set test character movement speed to 1.0 to avoid timeouts

skaldarnar committed:
test: skip "timeout tests" and increase timeout to 10s

skaldarnar committed:
test: pretty-print failing tests

skaldarnar commented:
Locally, there is a single failing test case for me now:

```
13:23:23.542 [Test worker] INFO  o.t.module.behaviors.MovementTests - movement plugin combination: [walking, leaping, falling]

Test character is not at target position.
  start   : ( 0.000E+0  4.100E+1  0.000E+0)
  target  : ( 5.000E+0  4.100E+1  0.000E+0)
  current : ( 0.000E+0  4.100E+1  0.000E+0)

  X    X|XX XX|XXXXX
 ==> expected: <( 5.000E+0  4.100E+1  0.000E+0)> but was: <( 0.000E+0  4.100E+1  0.000E+0)>
Expected :( 5.000E+0  4.100E+1  0.000E+0)
Actual   :( 0.000E+0  4.100E+1  0.000E+0)
```

All the other (positive) test cases pass nicely, usually in under 3 seconds per test case, summing up to a bit less than 2 minutes of execution time. The result is the same on CI, about 50% slower than running test locally with an execution time of 2min 40s.

jdrueckert committed:
test: add down and up again for WalkLeapFall combination

jdrueckert committed:
chore: disable failing tests until we fix the associated pathfinding

jdrueckert committed:
Merge branch 'develop' into fix/behavior-follow-ups

jdrueckert commented on assets/behaviors/common/doRandomMove.behavior:
```diff
@@ -1,9 +1,8 @@
-    set_target_nearby_block : { moveProbability: 65 }
-  },
-   move_to
+    { log: { message: "in doRandomMove Behavior"} },
```
do we want to keep this and similar log messages in behavior trees and also in code in for the time being and clean them up later or should we remove them before merging this so we don't forget about them?

skaldarnar replied:
I'll add some doc comment to `LogAction` and remove those debug statements.

skaldarnar replied:
See 34b61b8a40c8473fda73087bfdcfcb7ec92fe8ff

jdrueckert commented on assets/behaviors/creatures/critter.behavior:
```diff
@@ -1,25 +1,21 @@
+      ]
+    }
   ]
 }
```
```suggestion
}

```

jdrueckert committed:
Merge remote-tracking branch 'origin/develop' into fix/behavior-follow-ups

jdrueckert committed:
chore: reset MovementTests to develop state

jdrueckert committed:
chore: add empty line

jdrueckert committed:
Merge branch 'develop' into fix/behavior-follow-ups

skaldarnar committed:
doc: explain usage of LogAction; remove debug logs from behaviors

skaldarnar committed:
chore: remove unnecessary `sequence` in critter behavior

skaldarnar committed:
chore: improve TODO comment for debug rendering

skaldarnar commented on src/main/java/org/terasology/module/behaviors/actions/SetTargetToNearbyBlockNode.java:
```diff
@@ -27,47 +28,81 @@
             Vector3ic[] allowedBlocks = Iterators.toArray(
                     Iterators.transform(Iterators.filter(neighbors.iterator(),
-                                    (candidate) -> plugin.isReachable(currentBlock, candidate)),
+                                    (candidate) -> candidate.distanceSquared(startBlock) > currentDistance
```
This condition is probably not very good as it easily runs into local dead ends, but it at least prevents us from jumping back and forth while searching... We have a TODO below to improve on this.

skaldarnar committed:
chore: improve some TODO comments and docstrings

skaldarnar reviewed (commented):
Overall, this PR now does the following:

- style: format `.behavior` files
- chore: add TODO ideas and questions in several places
- chore: add some debug logging
- feature/fix: small adjustments to Behaviors logic

None of the remaining debug log statements does  any heavy computation, so it's fine to keep them in (for now) as they are not logged with the default configuration but might prove useful for debugging purposes.

LGTM
